src/native/libs/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -829,6 +897,16 @@ void* CryptoNative_SslGetData(SSL* ssl)
     return SSL_get_ex_data(ssl, 0);
 }
 
+int32_t CryptoNative_SslCtxSetData(SSL_CTX* ctx, void *ptr)
+{
+    return SSL_CTX_set_ex_data(ctx, 0, ptr);
+}
+
+void* CryptoNative_SslCtxGetData(SSL_CTX* ctx)
+{
+    return SSL_CTX_get_ex_data(ctx, 0);
```

Comment:
I'm a little concerned that we're just asserting an ex_data slot instead of going through the process of requesting a slot ID.

We don't claim exclusive control over the OpenSSL libraries, and if someone happens to be interacting with `SSL_CTX*`/`SSL*` themselves they may have registered an ex_data slot that has a cleanup routine registered.  Then they'd start receiving our GCHandle values in their cleanup routine, which will probably do all sorts of bad things.

We probably want to do something in our EnsureLibSslInitialized routine like

```C
g_ssl_ctx_sessioncache_index = SSL_CTX_get_ex_new_index(0, NULL, NULL, NULL, NULL);
```

(and a separate one for whatever we're storing in the `SSL*`)

Sure, it'll probably always be 0, but then we're at least being good cooperative users of the library.

Reply:
We could. But that generally means more allocations for each SslStream as far as I can tell. Somebody _can_ call OpenSSL directly and use some extension data. But there is no good way IMHO how they can get hand on our instances. e.g. they would need to get the SafeHandle from SslStream's PAL and fiddle with it directly via casting.  
I think it is ok to use same index on their instances while we attached our IntPtr on instances created by us. 

Any thought on this @stephentoub?

Reply:
>I think it is ok to use same index on their instances while we attached our IntPtr on instances created by us.

The concern I have is that someone registers and obtains index 0, and provides a deallocator routine to it.  When the `SSL` (or `SSL_CTX`, as appropriate) is freeing, then it goes through all registered indexes and calls the deallocator (if present).

So some process that happens to use an SSL/SSL_CTX on their own may all of a sudden start calling `free` (or whatever their custom deallocator is) and passing in our GCHandle value; because it has no idea which source created and populated slot 0.

>  But that generally means more allocations for each SslStream as far as I can tell

It won't have any impact on GC memory.

From my recollection of reading through the ex_data initialization any registered ID that has no registered pre-allocator just increases the cost of the for loop (it doesn't pre-allocate the memory behind the `STACK_OF`, that waits until the first call to set), so it won't really impact native memory either.  Just a couple extra instructions (walking the loop one time instead of 0, and checking that the slot-zero-preallocator is null).

Reply:
ok. I think the custom (de)allocators could be problem as registered per index. I will take another look.

Reply:
I did more testing and here is the outcome. I wrote this to have better insight:
```c
#include <openssl/ssl.h>
#include <stdio.h>

int log = 0;

int total = 0;
int count = 0;

void* mmalloc(size_t size, const char * location, int line)
{
    if (log) printf("malloc: %lu from %s:%d\n", size, location, line);

    total+=size;
    count++;
    return malloc(size);
}

void* mrealloc(void* ptr, size_t size, const char * location, int line)
{
    if (log) printf("realloc: %lu from %s:%d\n", size, location, line);
    return realloc(ptr, size);
}

void mfree(void* ptr, const char * location, int line)
{
    return free(ptr);
}

int main(int argc, char** argv)
{
    int idx = 0;
    int idx2 = 0;

    CRYPTO_set_mem_functions(mmalloc, mrealloc, mfree);

    SSL_library_init();
    OpenSSL_add_ssl_algorithms();
    SSL_load_error_strings();

    if (argc > 1)
    {
        for (int j = 0; j < 1; j++)
        {
            idx = SSL_CTX_get_ex_new_index(0, NULL, NULL, NULL, NULL);
            idx2 = SSL_get_ex_new_index(0, NULL, NULL, NULL, NULL);
        }
        printf("Using index of %d %d\n", idx, idx2);
    }

    log=1;

    for (int i = 0; i < 125; i++)
    {
        SSL_CTX* ctx = SSL_CTX_new(SSLv23_method());
        SSL_CTX_set_ex_data(ctx, idx, (void*)123456);
        SSL* ssl = SSL_new(ctx);
        SSL_set_ex_data(ssl, idx2, (void*)123456);
    }

    printf("Total %u allocated bytes on %d calls\n", total, count);

    return 0;
}
```

Interestingly, aside from registration the allocations are identical when I register extra index. When I increase number of extensions it would eventually kick in extra `realloc`. So we may be in luck for gap in alignment. (x64)
Number of allocated chunks does not increase with number of allocated `SSL_CTX` or `SSL` structures. That proofs the idea of allocating the `STACK_OF` as one blob. 

now, `SSL_CTX_get_ex_new_index` and `SSL_get_ex_new_index` start at 1 and then they increase. So it seems the legacy 0 e.g.`SSL_set_app_data` would never interfere with any new registration. And there is no need to worry about custom (de)allocator for that particular index. 

Since it seems very cheap, I don't mind to allocate private index for us. But I also feel that the current code is safe using the `0` index.

Reply:
Ah, yes, it looks like 0 is intentionally skipped because of SSL_set_app_data.  I missed that when I was looking at how ex_data worked (or forgot it).

Since the SSL_CTX and SSL pointers aren't accepted or exposed, and there's no deallocator on slot 0, I'm OK with using slot 0.

Thanks for doing the digging.